pmd-core/src/main/java/net/sourceforge/pmd/benchmark/TimeTracker.java

```diff
@@ -0,0 +1,296 @@
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A time tracker class to measure time spent on different sections of PMD analysis.
+ * The class is thread-aware, allowing to differentiate CPU and wall clock time.
+ * 
+ * @author Juan Martín Sotuyo Dodero
+ */
+public final class TimeTracker {
+
+    private static boolean trackTime = false;
+    private static long wallClockStartMillis = -1;
+    private static final ThreadLocal<Queue<TimerEntry>> TIMER_ENTRIES;
+    private static final ConcurrentMap<TimedOperationKey, TimedResult> ACCUMULATED_RESULTS = new ConcurrentHashMap<>();
+    private static final TimedOperation NOOP_TIMED_OPERATION = new TimedOperation() {
+
+        @Override
+        public void close() {
+            // noop
+        }
+
+        @Override
+        public void stop(final int count) {
+         // noop
+        }
+
+        @Override
+        public void stop() {
+         // noop
+        }
+    };
+    
+    static {
+        TIMER_ENTRIES = new ThreadLocal<Queue<TimerEntry>>() {
+            @Override
+            protected Queue<TimerEntry> initialValue() {
+                return Collections.asLifoQueue(new LinkedList<TimerEntry>());
+            }
+        };
+    }
+    
+    private TimeTracker() {
+        throw new AssertionError("Can't instantiate utility class");
+    }
+    
+    /**
+     * Starts global tracking. Allows tracking operations to take place and starts the wall clock.
+     * Must be called once PMD starts if tracking is desired, no tracking will be performed otherwise.
+     */
+    public static void startGlobalTracking() {
+        wallClockStartMillis = System.currentTimeMillis();
+        trackTime = true;
+        ACCUMULATED_RESULTS.clear(); // just in case
+        initThread(); // init main thread
+    }
+    
+    /**
+     * Stops global tracking. Stops the wall clock. All further operations will be treated as NOOP.
+     * @return The timed data obtained through the run.
+     */
+    public static TimingReport stopGlobalTracking() {
+        if (!trackTime) {
+            return null;
+        }
+        
+        finishThread(); // finish the main thread
+        trackTime = false;
+        
+        // Fix UNACCOUNTED metric (total time is meaningless as is call count)
+        final TimedResult unaccountedResult = ACCUMULATED_RESULTS.get(
+                new TimedOperationKey(TimedOperationCategory.UNACCOUNTED, null));
+        unaccountedResult.totalTime.set(unaccountedResult.selfTime.get());
+        unaccountedResult.callCount.set(0);
+        
+        return new TimingReport(System.currentTimeMillis() - wallClockStartMillis, ACCUMULATED_RESULTS);
+    }
+    
+    /**
+     * Initialize a thread, starting to track it's own time.
+     */
+    public static TimedOperation initThread() {
```

Comment:
None of the callers seem to use the return from this method.  Is that okay? Is it needed?